File: source/Scanner.java
import java.util.*;

public class Scanner {
    
    private String source;
    private final List<Token> tokens = new ArrayList<Token>();
    private int start = 0, current = 0, line = 1;

    private final static Map<String, TokenType> keywords = new HashMap<String, TokenType>();
    static {
        keywords.put("class", TokenType.CLASS);
    }

    public Scanner(String source) {
        this.source = source;
    }

    public List<Token> scanTokens() {
        while (!isAtEnd()) {
            start = current;
            scanToken();
        }
        tokens.add(new Token(TokenType.EOF, "", null, line));
        return tokens;
    }

    private void scanToken() {
        char c = advance();
        switch (c) {
            case '(': addToken(TokenType.LEFT_PAREN); break;
            case ')': addToken(TokenType.RIGHT_PAREN); break;
            case '{': addToken(TokenType.LEFT_BRACE); break;
            case '}': addToken(TokenType.RIGHT_BRACE); break;
            case '+': addToken(TokenType.ADD); break;
        
            default:
                if (isDigit(c)) {
                    number();
                } else if (isAlpha(c)) {
                    identifier();
                    break;
                }
                break;
        }
    }

    private void addToken(TokenType type) {
        addToken(type, null);
    }

    private void addToken(TokenType type, Object literal) {
        String text = source.substring(start, current);
        tokens.add(new Token(type, text, literal, line));
    }

    private boolean isAlpha(char c) {
        return 
        (c >= 'a' && c <= 'z') || 
        (c >= 'A' && c <= 'Z') ||
        (c == '_');
    }

    private boolean isAlphaNumeric(char c) {
        return isAlpha(c) || isDigit(c);
    }

    private boolean isDigit(char c) {
        return c >= '0' && c <= '9';
    }

    public void identifier() {
        while (isAlphaNumeric(peek())) advance();
        String literal = source.substring(start, current);
        boolean isKeyword = keywords.containsKey(literal);
        if (isKeyword) {
            addToken(keywords.get(literal));
        } else {
            addToken(TokenType.IDENTIFIER, literal);
        }
    }
    
    public void number() {
        while (isDigit(peek())) advance();
        if (match('.')) {
            while (isDigit(peek())) advance();
        }
        addToken(TokenType.NUMBER, Double.parseDouble(source.substring(start, current)));
    }

    private char advance() {
        current++;
        return source.charAt(current-1);
    }

    private char peek() {
        if (isAtEnd()) return '\0';
        return source.charAt(current);
    }

    private char peekNext() {
        if (current >= source.length() - 1) return '\0';
        return source.charAt(current + 1);
    }

    private boolean match(char expected) {
        if (isAtEnd()) return false;
        if (source.charAt(current) != expected) return false;
        current++;
        return true;
    }

    private boolean isAtEnd() {
        return current >= source.length();
    }
}

File: source/Scanner.kt
import java.util.ArrayList
import java.util.HashMap

class Scanner(private val source: String) {
    private val tokens: MutableList<Token> = ArrayList()
    private var start = 0
    private var current = 0
    private var line = 1

    companion object {
        private val keywords: MutableMap<String, TokenType> = HashMap()

        init {
            keywords["class"] = TokenType.CLASS
            keywords["print"] = TokenType.PRINT
            keywords["true"] = TokenType.TRUE
            keywords["false"] = TokenType.FALSE
            keywords["var"] = TokenType.VAR
            keywords["if"] = TokenType.IF
            keywords["else"] = TokenType.ELSE
            keywords["and"] = TokenType.AND
            keywords["or"] = TokenType.OR
            keywords["while"] = TokenType.WHILE
            keywords["function"] = TokenType.FUNCTION
        }
    }

    fun scanTokens(): List<Token> {
        while (!isAtEnd) {
            start = current
            scanToken()
        }
        tokens.add(Token(TokenType.EOF, "", null, line))
        return tokens
    }

    private fun scanToken() {
        when (val c = advance()) {
            '(' -> addToken(TokenType.LEFT_PAREN)
            ')' -> addToken(TokenType.RIGHT_PAREN)
            '{' -> addToken(TokenType.LEFT_BRACE)
            '}' -> addToken(TokenType.RIGHT_BRACE)
            '+' -> addToken(TokenType.PLUS)
            '-' -> addToken(TokenType.MINUS)
            '*' -> addToken(TokenType.STAR)
            '!' -> addToken(if (match('=')) TokenType.BANG_EQUAL else TokenType.BANG)
            '=' -> when {
                    match('=') -> addToken(TokenType.EQUAL_EQUAL)
                    match('>') -> addToken(TokenType.ARROW)
                    else -> addToken(TokenType.EQUAL)
                }

            '>' -> addToken(if (match('=')) TokenType.GREATER_EQUAL else TokenType.GREATER)
            '<' -> addToken(if (match('=')) TokenType.LESS_EQUAL else TokenType.LESS)
            ';' -> addToken(TokenType.SEMICOLON)
            ',' -> addToken(TokenType.COMMA)
            '/' -> {
                if (match('/')) {
                    while (peek() != '\n' && !isAtEnd) advance()
                } else {
                    addToken(TokenType.SLASH)
                }
            }
            '"' -> string()
            else -> if (isDigit(c)) {
                number()
            } else if (isAlpha(c)) {
                identifier()
            }
        }
    }

    private fun addToken(type: TokenType?, literal: Any? = null) {
        val text = source.substring(start, current)
        type?.let { Token(it, text, literal, line) }?.let { tokens.add(it) }
    }

    private fun isAlpha(c: Char): Boolean {
        return c >= 'a' && c <= 'z' ||
                c >= 'A' && c <= 'Z' ||
                c == '_'
    }

    private fun isAlphaNumeric(c: Char): Boolean {
        return isAlpha(c) || isDigit(c)
    }

    private fun isDigit(c: Char): Boolean {
        return c >= '0' && c <= '9'
    }

    fun identifier() {
        while (isAlphaNumeric(peek())) advance()
        val literal = source.substring(start, current)
        val isKeyword = keywords.containsKey(literal)
        if (isKeyword) {
            addToken(keywords[literal])
        } else {
            addToken(TokenType.IDENTIFIER, literal)
        }
    }

    fun number() {
        while (isDigit(peek())) advance()
        if (match('.')) {
            while (isDigit(peek())) advance()
        }
        addToken(TokenType.NUMBER, source.substring(start, current).toDouble())
    }

    fun string() {
        while (peek() != '"' && !isAtEnd) {
            if (peek() == '\n') line++
            advance()
        }

        if (isAtEnd) {
            error("Unterminated string")
            return
        }
        advance()
        val value = source.substring(start+1, current-1)
        addToken(TokenType.STRING, value)
    }

    private fun advance(): Char {
        current++
        return source[current - 1]
    }

    private fun peek(): Char {
        return if (isAtEnd) '\u0000' else source[current]
    }

    private fun peekNext(): Char {
        return if (current >= source.length - 1) '\u0000' else source[current + 1]
    }

    private fun match(expected: Char): Boolean {
        if (isAtEnd) return false
        if (source[current] != expected) return false
        current++
        return true
    }

    private val isAtEnd: Boolean
        private get() = current >= source.length
}